### ISSUE TITLE
Pay with Amazon JS Loads on Every Page

### ISSUE THREAD

**Issue by islandskater43** [amzn/amazon-payments-magento-2-plugin#29]
The Pay with Amazon JS loads on every page, regardless of whether a button or widget is rendered on a page. This increases total page load time by a few seconds on crucial pages (home page, product detail page, etc). 

Can we investigate what options are available to suppress loading widgets.js on pages unnecessarily?

**Comment by gewaechshaus** [comment]
@islandskater43 - we noticed that to... I should be partially needed for the minicart, but it's a question of architecture and initialisation.

**Comment by scuf-davidg** [comment]
We'd like to see this as well.  We don't need 34 extra files loaded on every page.
To note, we do not use the minicart at all and we only show payment abilities in the checkout process.

**Comment by andidhouse** [comment]
Hi there,

is this solved. We are ready to start but implementing a payment extension which implements files on every page is nonsense as most of the shops do only need payment information in the checkout.

Can someone pls give a timeline here as for us it is a big blocker which makes the extension not usable at the moment.

Many thanks!

**Comment by danielneu** [comment]
Hi,

this will be part of version 1.1.1 and is already merged into the master.
To give some context, I would not rate this nonsense:
Amazon Pay is an improved checkout and as an express option the entry point is not at the end of the checkout but everywhere a checkout can be started.
This includes e.g. the mini-cart which is available on any page.

Best wishes 
Daniel

**Comment by andidhouse** [comment]
Hi Daniel,

without an option to completely deactivate the integration on all other sites then the checkout pages is in my opinion "nonsense" and i can tell you why.

- Site speed decreases also for users which do not use your mentioned option to implement your payment for example at product page or mini-cart --> bad for user experience and SEO

- But most important think about 3-5 implemented payment extensions (which is quite normal for an commerce site) which do the same as your extension --> conflicts, massive more files to download on every page etc etc...

I think this option should really be optional and it should be possible to completely deactivate this at all because most store owners will not use this at all.

I know for amazon it would be great to have their payment most prominent in every shop but this is - i think - a complete wrong way of doing it.

Thanks,

André

**Comment by scuf-davidg** [comment]
@danielneu 

> this will be part of version 1.1.1 and is already merged into the master.

Are you saying we can test this out to see if it's no longer loading the extra files on all the pages?

**Comment by andidhouse** [comment]
It would really be great to know when 1.1.1 is expected to be released as a stable version. Is there any timeline?

Thanks!

**Comment by danielneu** [comment]
My bad, this is assigned to the 1.1.2 milestone, I confused things - sincere apologies.
We are working on this release at this time.

Version 1.1.1 was released already: https://github.com/amzn/amazon-payments-magento-2-plugin/releases
Respectively: https://marketplace.magento.com/amzn-amazon-pay-and-login-magento-2-module.html

Best wishes 
Daniel

**Comment by scuf-davidg** [comment]
@danielneu Please work on getting this merged so we can test.

**Comment by andidhouse** [comment]
@danielneu can you pls give us a timeline here. We are ready to go with m2 and amazon pay with a view shops but not with a payment method which implements scripts on every site.

If there is no timeline or info can can not rely on this payment option for the shops we launch at the moment and will go for PayPal plus instead.

Thanks

**Comment by danielneu** [comment]
Follow up on this was done today. Functionality is given and accepted. Verification from your side can be done using #71 already.

Smaller changes + translations have been requested.

This will definitely be part of the next release, we will handle this with priority.

**Comment by danielneu** [comment]
This was merged now and is part of master. Documentation update will be up in a couple of minutes.

Resolving, watch out for the new release soon.

**Comment by scuf-davidg** [comment]
@danielneu  Could you do the version bump so this will be upgradeable by composer and magento marketplace?

**Comment by danielneu** [comment]
Hi @scuf-davidg,

once testing is completed and we do the release of 1.1.2, yes.
This is hopefully happening this week.

If you want to go ahead (e.g. for testing on staging), you can update using composer by providing "dev-master" as the version.

We only update versions and cut a release once QA is complete.

Best wishes 
Daniel

**Comment by danielneu** [comment]
Version 1.1.2, including this feature, was released just now.
Marketplace update still pending, documentation updated.

Options > Advanced > Frontend allows you to disable the Amazon Pay button in various locations, which will then also not load our JavaScript code.

**Comment by scuf-davidg** [comment]
From what I see, this did not fix the issue.  Still getting amazon js loading on our homepage and other pages.

Reviewing the code, this doesn't prevent Amazon from loading the js files at all, just stopping the button from displaying in the minicart.  This doesn't fix the initial issue reported.

**Comment by danielneu** [comment]
@scuf-davidg 

if you disable the mini cart button and the button in the authentication popup, it will.
The only reason the file is loaded on all pages is the fact the mini cart and authentication popup is loaded on all pages.

Best wishes 
Daniel

**Comment by andidhouse** [comment]
HI @danielneu do i unterstand the logic right:

- you have to deactivate in the extension not to show on minicart/any page AND you also need to deactivate the minicart function itself?

If so - makes no sense to me. What about people who like the minicart function but do not want to show the amazon pay here.

In my mind if the setting in the extension backend is set to "just show on checkout - it should in any case - no matter how - just be shown in the checkout.

Think about PayPal and other payment services do the same - 5 buttons in the minicart then?

**Comment by danielneu** [comment]
I'm talking about this and the other options mentioned there: 
https://amzn.github.io/amazon-payments-magento-2-plugin/configuration.html#amazon-pay-button-is-visible-on-minicart

Best wishes 
Daniel

**Comment by scuf-davidg** [comment]
Still doesn't fix our issue. 

Does this somehow work different with Varnish?  Was it tested with Varnish enabled?

We do not have minicart enabled anywhere and both options needed are turned off for this to not load on every page.
